core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -61,7 +62,12 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
     this(sc, new SystemClock)
   }
 
-  sc.addSparkListener(this)
+  if (clock.isInstanceOf[SystemClock]) {
```

Comment:
This is kind of a weird check; I'd favor either some explicit "testMode" boolean parameter to the constructor, or, my preferred, making the test ignore executors that match `SparkContext.DRIVER_IDENTIFIER`.